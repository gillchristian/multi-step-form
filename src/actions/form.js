import { createAction } from 'redux-actions'

import {
  SUBMIT_FORM_BEGIN,
  SUBMIT_FORM_END,
  VALIDATE_FIELD_BEGIN,
  VALIDATE_FIELD_END,
  ON_FIELD_CHANGE,
  MOVE_TO_NEXT_STEP,
} from '../constants/action-types'

// FSA :: Flux Standard Action @link https://goo.gl/Tdu4BA

// submitForm :: () -> FSA
export const submitForm = createAction(SUBMIT_FORM_BEGIN)
// submitFormEnd :: a -> FSA
export const submitFormEnd = createAction(SUBMIT_FORM_END)

// validateAsyncBegin :: String -> (() -> FSA)
export const validateAsyncBegin =
  field => createAction(VALIDATE_FIELD_BEGIN, () => {}, () => field)

// validateAsyncEnd :: String -> (a -> FSA)
export const validateAsyncEnd =
  field => createAction(VALIDATE_FIELD_END, p => p, () => field)

// moveToNextStep :: () -> FSA
export const moveToNextStep = createAction(MOVE_TO_NEXT_STEP)

// onFieldChange :: String -> String -> FSA
export const onFieldChange = createAction(
  ON_FIELD_CHANGE,
  (field, value) => ({value}),
  (field) => field
)

