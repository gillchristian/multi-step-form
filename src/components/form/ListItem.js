import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { Margin, Text } from '../styled'

const ListItemWrapper = styled.div`
  display: flex;
  align-items: center;

  margin-bottom: 10px;
`

function ListItem({children, label}) {
  return (
    <ListItemWrapper>
      <Margin right="10px">
        {children}
      </Margin>
      { Boolean(label) && <Text>{label}</Text> }
    </ListItemWrapper>
  )
}

ListItem.propTypes = {
  children: PropTypes.node,
  label: PropTypes.string,
}

export default ListItem
