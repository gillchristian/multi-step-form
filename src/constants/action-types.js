export const MOVE_TO_NEXT_STEP = 'form:steps:move-to-next'

export const ON_FIELD_CHANGE = 'form:data:update-field'

export const VALIDATE_FIELD_BEGIN = 'form:validation:async-begin'
export const VALIDATE_FIELD_END = 'form:validation:async-end'

export const SUBMIT_FORM_BEGIN = 'form:submition:begin'
export const SUBMIT_FORM_END = 'form:submition:end'

