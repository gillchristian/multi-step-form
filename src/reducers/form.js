import { combineReducers } from 'redux'

import steps from './steps'
import data from './data'
import validation from './validation'
import submition from './submition'

export default combineReducers({
  steps,
  data,
  validation,
  submition,
})

