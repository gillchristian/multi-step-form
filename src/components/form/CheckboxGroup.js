import React from 'react'
import PropTypes from 'prop-types'

import { CHECKABLE_ITEM_LIST } from '../../constants/prop-types'
import { Checkbox } from '../styled'
import ListItem from './ListItem'

function CheckBoxGroup({items, onChange, value, multiple}) {

  const isChecked = (v) => multiple ? value.includes(v) : value === v

  const onClick = (v) => isChecked(v)
    ? onChange(multiple ? value.filter(item => item !== v) : v)
    : onChange(multiple ? [...value, v] : v)

  return (
    <div>
      {
        items.map((item) => (
          <ListItem label={item.label} key={item.value}>
            <Checkbox
              checked={isChecked(item.value)}
              onClick={() => onClick(item.value)}
            />
          </ListItem>
        ))
      }
    </div>
  )
}

CheckBoxGroup.propTypes = {
  items: CHECKABLE_ITEM_LIST.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.any.isRequired,
  multiple: PropTypes.bool,
}

export default CheckBoxGroup

