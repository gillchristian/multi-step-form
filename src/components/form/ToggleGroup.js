import React from 'react'
import PropTypes from 'prop-types'

import { CHECKABLE_ITEM_LIST } from '../../constants/prop-types'
import { Toggle } from '../styled'
import ListItem from './ListItem'

function ToggleGroup({items, onChange, value}) {
  return (
    <div>
      {
        items.map((item) => (
          <ListItem label={item.label} key={item.value}>
            <Toggle
              checked={value === item.value}
              onClick={() => onChange(item.value)}
            />
          </ListItem>
        ))
      }
    </div>
  )
}

ToggleGroup.propTypes = {
  items: CHECKABLE_ITEM_LIST.isRequired,
  value: PropTypes.any.isRequired,
  onChange: PropTypes.func.isRequired,
}

export default ToggleGroup

