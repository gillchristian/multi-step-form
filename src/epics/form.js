import R from 'ramda'
import { Observable } from 'rxjs'

import {
  ON_FIELD_CHANGE,
  VALIDATE_FIELD_BEGIN,
  VALIDATE_FIELD_END,
  SUBMIT_FORM_BEGIN,
} from '../constants/action-types'
import * as api from '../api/index'
import { submitFormEnd,
  moveToNextStep,
  validateAsyncEnd,
} from '../actions/form'
import { getFieldValue, getFormData } from '../reducers/data'
import { isCurrentStep } from '../reducers/steps'

// checkIt :: State -> String -> Observable(Promise)
const checkIt = R.compose(
  Observable.fromPromise,
  api.checkIt,
  getFieldValue
)

// submitIt :: State -> Observable(Promise)
const submitIt = R.compose(
  Observable.fromPromise,
  api.submitIt,
  getFormData
)

// handleCheckError :: Error -> Observable(FSA)
const handleCheckError = R.compose(
  Observable.of,
  validateAsyncEnd('text')
)

// handleSubmitError :: Error -> Observable(FSA)
const handleSubmitError = R.compose(
  Observable.of,
  submitFormEnd
)

// syncTypes :: FSA -> Bool
const syncTypes = ({ type, error, meta }) =>
  (type === ON_FIELD_CHANGE && meta !== 'text') ||
  (type === VALIDATE_FIELD_END && !error)

// nextStepEpic handles transition to the next step in the form
export const nextStepEpic = (action$, { getState }) => action$
  .filter(syncTypes)
  .filter(({ meta }) => isCurrentStep(getState(), meta))
  .mapTo(moveToNextStep())

// asyncValidationEpic handles async validation of the text field
export const asyncValidationEpic = (action$, { getState }) => action$
  .ofType(VALIDATE_FIELD_BEGIN)
  .filter(R.compose(R.equals('text'), R.prop('meta')))
  .mergeMap(
    () => checkIt(getState(), 'text')
      .map(validateAsyncEnd('text'))
      .catch(handleCheckError)
  )

// syncValidationEpic handles sync validation whenever a field changes
export const syncValidationEpic = (action$) => action$
  .ofType(ON_FIELD_CHANGE)
  .filter(({ meta }) => meta !== 'text')
  .map(R.ifElse(
    R.compose(R.isEmpty, R.path(['payload', 'value' ])),
    ({meta}) => validateAsyncEnd(meta)(new Error('Please pick one')),
    ({meta}) => validateAsyncEnd(meta)()
  ))

// submitEpic handles form submition
export const submitEpic = (action$, { getState }) => action$
  .ofType(SUBMIT_FORM_BEGIN)
  .mergeMap(
    () => submitIt(getState())
      .mapTo(submitFormEnd())
      .catch(handleSubmitError)
  )

