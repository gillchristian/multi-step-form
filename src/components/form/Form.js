import React from 'react'
import PropTypes from 'prop-types'

import { Input, Button, Spinner, Help, Progress } from '../styled'
import CheckboxGroup from './CheckboxGroup'
import ToggleGroup from './ToggleGroup'
import Section from '../Section'

function Form({
  isSubmitting,
  didSubmit,
  message,
  didSubmitionFailed,
  total,
  completed,
  isFinished,
  isPristine,
  isValidating,
  fieldValue,
  hasError,
  fieldError,
  shouldDisplayStep,
  onFieldChange,
  validateText,
  handleSubmit,
}) {

  const aItems = [
    { value: 'a1', label: 'Cheese' },
    { value: 'a2', label: 'BBQ' },
  ]

  const bItems = [
    { value: 'b1', label: 'Dog' },
    { value: 'b2', label: 'Cat' },
  ]

  const cItems = [
    { value: 'c1', label: 'Backend' },
    { value: 'c2', label: 'Frontend' },
    { value: 'c3', label: 'Fullstack' },
  ]

  return (
    <form onSubmit={handleSubmit}>

      <Progress
        total={total}
        count={completed}
        success={didSubmit && !didSubmitionFailed}
        error={didSubmit && didSubmitionFailed}
      />

      {
        shouldDisplayStep('a') &&
        <Section title="Burgers with?" bg="darkseagreen">
          <CheckboxGroup
            items={aItems}
            value={fieldValue('a')}
            onChange={v => onFieldChange('a', v)}
            multiple
          />
          {
            hasError('a') && !isPristine('a') &&
            <Help type="danger">{fieldError('a')}</Help>
          }
        </Section>
      }

      {
        shouldDisplayStep('b') &&
        <Section title="Best buddy" bg="lightsteelblue">
          <ToggleGroup
            items={bItems}
            value={fieldValue('b')}
            onChange={v => onFieldChange('b', v)}
          />
          {
            hasError('b') && !isPristine('b') &&
            <Help type="danger">{fieldError('b')}</Help>
          }
        </Section>
      }

      {
        shouldDisplayStep('text') &&
        <Section title="Your at handle" bg="papayawhip">
          <Input
            placeholder="@asdf"
            value={fieldValue('text')}
            onChange={e => onFieldChange('text', e.target.value)}
            error={hasError('text') && !isPristine('text')}
          />
          {
            !isPristine('text') &&
            <Help type={hasError('text') ? 'danger' : 'success'}>
              {
                hasError('text')
                  ? fieldError('text')
                  : 'Yay!!!'
              }
            </Help>
          }
          {
            isValidating('text')
              ? <Button loading><Spinner /></Button>
              : <Button
                  error={hasError('text') && !isPristine('text')}
                  onClick={validateText}
                >
                  Check
                </Button>
          }
        </Section>
      }

      {
        shouldDisplayStep('c') &&
        <Section title="Role" bg="palevioletred">
          <CheckboxGroup
            items={cItems}
            value={fieldValue('c')}
            onChange={v => onFieldChange('c', v)}
          />
          {
            hasError('c') && !isPristine('c') &&
            <Help type="danger">{fieldError('c')}</Help>
          }
        </Section>
      }

      {
        isFinished &&
        <Section bg="rosybrown">
          {
            isSubmitting
              ? <Button loading><Spinner /></Button>
              : <Button type="submit" onClick={handleSubmit}>Submit</Button>
          }
          {
            didSubmit && !isSubmitting &&
            <Help type={didSubmitionFailed ? 'danger' : 'success'}>
              {message}
            </Help>
          }
        </Section>
      }

    </form>
  )
}

Form.propTypes = {
  isSubmitting: PropTypes.bool,
  didSubmit: PropTypes.bool,
  message: PropTypes.string,
  didSubmitionFailed: PropTypes.bool,
  total: PropTypes.number,
  completed: PropTypes.number,
  isFinished: PropTypes.bool,
  isPristine: PropTypes.func.isRequired,
  isValidating: PropTypes.func.isRequired,
  fieldValue: PropTypes.func.isRequired,
  hasError: PropTypes.func.isRequired,
  fieldError: PropTypes.func.isRequired,
  shouldDisplayStep: PropTypes.func.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  validateText: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
}

export default Form
