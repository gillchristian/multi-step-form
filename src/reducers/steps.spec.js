import { MOVE_TO_NEXT_STEP } from '../constants/action-types'
import { moveToNextStep } from '../actions/form'
import reducer, {
  INITIAL_STATE,
  isFinished,
  isCurrentStep,
  stepsCount,
  completedSteps,
} from './steps'

Object.freeze(INITIAL_STATE)

describe('form steps reducer', () => {

  describe(MOVE_TO_NEXT_STEP, () => {
    it('moves to the next step', () => {
      const state = reducer(INITIAL_STATE, moveToNextStep())

      expect(state).toMatchSnapshot()
    })

    it('does nothing if already in the last stpe', () => {
      const prevState = {
        current: 'submit',
        next: [],
        prev: [ 'a', 'b', 'text', 'c' ],
      }

      const state = reducer(prevState, moveToNextStep())

      expect(state).toEqual(prevState)
    })
  })

})

describe('form steps selectors', () => {
  const state = { form: { steps: INITIAL_STATE } }

  describe('isFinished', () => {
    it('check if the current step is submit', () => {
      expect(isFinished(state)).toBe(false)
      expect(isFinished({ form: { steps: { current: 'submit' } } })).toBe(true)
    })
  })

  describe('isCurrentStep', () => {
    it('check if the provided step is the current', () => {
      expect(isCurrentStep(state, 'a')).toBe(true)
      expect(isCurrentStep(state, 'b')).toBe(false)
    })
  })

  describe('stepsCount', () => {
    it('calculates the total amount of steps', () => {
      expect(stepsCount(state)).toBe(4)
    })
  })

  describe('completedSteps', () => {
    it('calculates the amount of completed steps', () => {
      expect(completedSteps(state)).toBe(0)

      const _state = {
        form: {
          steps: { current: 'text', prev: ['a', 'b'], next: ['c', 'submit'] },
        },
      }
      expect(completedSteps(_state)).toBe(2)
    })
  })

})
