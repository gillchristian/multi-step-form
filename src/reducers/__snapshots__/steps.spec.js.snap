exports[`form steps reducer form:steps:move-to-next moves to the next step 1`] = `
Object {
  "current": "b",
  "next": Array [
    "text",
    "c",
    "submit",
  ],
  "prev": Array [
    "a",
  ],
}
`;
